packages/libs/avector/src/lib.rs

```diff
@@ -0,0 +1,551 @@
+//! Concurrent Read Optimized Vector
+//!
+//! Vector which allows concurrent read access, simultaneous to writing.
+//! The vector can only be manipulated through pushing of items, once an item is pushed it can no
+//! longer be accessed mutably.
+//!
+//! While reading is concurrent and can happen while values are being written,
+//! values can only be written synchronously and are guarded by a spinlock.
+//! This means that [`AVec`] works best when contention is low and values are mostly being
+//! read.
+//!
+//! ## How it works
+//!
+//! [`AVec`] is based on two parts: length (through an atomic) and an intrusive list of
+//! arrays which store items.
+//! The size of a bucket can be controlled through the const generic `N`, allowing one to tradeoff
+//! speed vs. allocated space.
+//!
+//! The inner workings are best explained through a demonstration:
+//!
+//! ```text
+//! Step 1:
+//!     Vec<u8, 4>: [_, _, _, _]
+//!     A: Acquire Lock
+//!     B: Read, Length: 0
+//!
+//! Step 2:
+//!     Vec<u8, 4>: [_, _, _, _]
+//!     A: Load Length (0)
+//!     B: .next(): None
+//!
+//! Step 3:
+//!     Vec<u8, 4>: [1, _, _, _]
+//!     A: Push Value
+//!     B: Read, Length: 0
```

Comment:
Why the length is `0` at this point?

Reply:
I will redo this section, in hindsight my attempt at explaining the inner workings failed miserably here.

Reply:
I will resolve these comments then for now and revisit this section later.